### DISCUSSION TITLE
Independent and Optimized dataloader to the teacher and student model

### DISCUSSION THREAD

**Discussion by AhmedHussKhalifa** [yoshitomo-matsubara/torchdistill#167] [Q&A] [answered]
Hey,

I am trying to tailor your framework to accommodate my experiments on imagenet. One of my experiments is to implement this paper [Knowledge distillation: A good teacher is patient and consistent](https://arxiv.org/abs/2106.05237) where they do some data augmentation at the teacher and student side. Each model has different input data. I expect I  would need to creat each of them a different dataloader. I will also maintain the same augmentation method for each image to be able to store the output vector of the teacher using [default_idx2subpath](https://github.com/yoshitomo-matsubara/torchdistill/blob/f7e757f63982c76bc3e9db4354edf1e008963f8a/torchdistill/datasets/wrapper.py#L14) function on my hard drive (SSD). I figured that saving the data on the SSD compared to loading the images made the training runs slower. 

I am trying to train a Student model -ResNet18- using 2 GPUs on ImageNet and a teacher model -Resnet34-. Could u recommend what is the best scenario for this pipeline to run the code faster giving the current resources?

**Comment by yoshitomo-matsubara** [comment, marked as answer]
Hi @AhmedHussKhalifa 
If you need two different data loaders in training session for each of teacher and student models, torchdistill would not be a good option.
A potential solution is to prepare two different transform functions given an input sample and use the transform pipelines as part of teacher and student models respectively, which could be done with torchdistill.

Also, if the cached vectors from your teacher model is relatively big, it may not speed-up your training. So I would suggest using distributed training mode with multiple GPUs instead.
